### DISCUSSION TITLE
Return joined record on different key.

### DISCUSSION THREAD

**Discussion by apstanisic** [directus/directus#3633] [Feature Requests]
Allow option to change where joined record is shown.

For example if we have
```ts
interface Post {id: string; user_id: string; }
interface User {id: string; name: string; }
```
When we request `items/posts?fields=*,user_id.*` we will get 
```ts 
interface ResultData { id: string; user_id: User };
```
It would be cool if we could write something like: `items/posts?fields=*,user_id>user.*` that will return this:
```ts
interface ResultData { id: string; user_id: string; user: User };
```
That would join but it would set joined data to different field.

Most of the languages don't have union types, so we can't do something equivalent in TypeScript:
```ts
interface A { b: string | B; }
```

**Comment by philleepflorence** [comment]
Are you trying to request a feature similar to MySQL Alias?
```
SELECT 
   [column_1 | expression] AS descriptive_name
FROM table_name;
```
Select a field under a different name?

**Reply by apstanisic** [reply]
Not quite, `AS` returns only one column with different name. What I'm thinking off is all columns that are joined should be object with different key.
For example:
We have posts table `id | title | user_id` and users table `id | first_name `
Currently if you request `*`, it will return something like this
```json
{"id": 1, "title": "My title", "user_id": "some-uuid" }
```
and if you request `*,user_id.*`, it will return this:
```json
{"id": 1, "title": "My title", "user_id": { "id": "some-uuid", "first_name": "Test Name" }  }
```
What I am thinking of is that by requesting something like this `*,user_id>user.*` that it returns something like this:
```json
{"id": 1, "title": "My title", "user_id": "some-uuid", "user": { "id": "some-uuid", "first_name": "Test Name" }  }
```
It will add additional field to json response with name that we specified with `>user`.
If we wanted field to be named `owner` we would pass `user_id>owner.*`.
That would still return `user_id` as string, and create additional field that would user data

**Comment by philleepflorence** [comment]
Sorry, maybe I was not clear, but what you explained is a feature of MySQL Alias...
```
SELECT user_id, user_id.* AS user...
```
I am trying to match what you mean to the feature of either a DB or the library used by Directus to communicate with all the DB types currently supported...

So instead of focusing only on the FK fields, if enacted, the feature could be used on any field.

Currently, in my extended version of D8, this is handled via a custom hook and helper that process most JSON responses.

I have it in mind to move the functionality to the core via parameters and building on the features available already in D9 should time and resources permit. Almost all my projects are data intensive so I do process or concat almost all responses for the web applications.

**Reply by apstanisic** [reply]
I know knex returns flat data, but I don't know how it then process joins for every db type and how Directus does nesting. 
Is `user_id.* as user` pseudo code or valid MySQL? If it's valid, than that's what I meant.
Can you show me hook and helper? I found repo, but I didin't find code that does this.

You're right. It can be a hook that does this. Something like:
```ts
  Object.entries(data).forEach(([key, val]) => {
    if (key.endsWith('_id') && typeof val === 'object' && !Array.isArray(val)) {
      data[key.substr(0, key.length - 3)] = val;
      data[key] = val.id;
    }
  });
```
But this only works if db follows convention that FK ends with `_id` and PK is `id`.

I've been working on SDK for Flutter/Dart and my main use case would be to be able to join with full type safety.
Dart does not have union types, so for example bellow, there is no way to set parent as both `String` and `DirectusFolder`.
I can only set it as `dynamic` and lose all type safety.
```dart
class DirectusFolder {
  String? id;
  String? name;
  String? parent;

  DirectusFolder({
    this.id,
    this.name,
    this.parent,
  });
}
```
Ideally it would be something like:
```dart
class DirectusFolder {
  String? id;
  String? name;
  String? parentId;
  DirectusFolder? parent;

  DirectusFolder({
    this.id,
    this.name,
    this.parent,
    this.parentId,
  });
}
```

**Reply by philleepflorence** [reply]
If you are using D9, you would need to translate the logic to Node, if it matches, I don't plan on doing that until D9 is post production and documentation somewhat commensurate.

The repo is: https://github.com/philleepflorence/directus-8-extensions
The helper is: https://github.com/philleepflorence/directus-8-extensions/blob/master/helpers/collections.php
The hook is: https://github.com/philleepflorence/directus-8-extensions/blob/master/hooks/response/hooks.php
The method is compile...also the helper files and methods tend to match the endpoint.

But, in your case, I think you will have to standardize the naming of fields and collections to be able to create a program to process the hook in a more Object Oriented Way. Directus has a pretty understandable formatting of fields, and all internal collections start with **directus_**.

I know I had to standardize the formats of the fields and collections when creating many automated scripts to support some DBs with 60+ collections, since I wanted to make the process of maintaining the DBs viable. 
E.G. I use the prefixes **app_**, **contents_**, and **join_**, to differentiate collections from those with **directus_**, to standardize and automate permissions et al, while also making it easy for admins to understand.

**Comment by rijkvanzanten** [comment]
Sounds like an interesting idea, but what is the use case for it? What's wrong with nesting it in the key? You could also name your m2o `user` instead of `user_id` if the `_id` part is what's confusing 🤔

**Reply by philleepflorence** [reply]
The reason I wanted to find an interpretation for the concept was to be able to discuss and understand it more. Also to limit the loss in translation.

There are multiple use cases for **Alias** when running a Select Query, while dealing with structured and relational data, it is something SQL can do faster than almost any processing language. One of the reasons I write so many extensions is because the projects I develop are more complex than a typical website, but the data must be structured in a format a non technical human can CRUD. 

And, for me to explain or understand something, it helps to find a reference if it exists.

As knex is the current library used for D9, it would make it easier to translate. validate, and standardize parameters into more complex queries depending on the support of the DB, than trying to reimagine a whole new concept,  especially if the tools already exists.

Imagine supporting functions like SUM, LENGTH, MIN, or the most common functions used in Excel or Spreadsheets, one would no longer need to return JSON from the DB and then process in another language. The data will return what you need. Admins can then write complex queries like they do in spreadsheets, like analytics or something.

**Advantages of MySQL Aliases**
It can make the field names more readable. 
It is useful when you use functions or procedures in the query - this is something I use a lot, like calculations, returning the calculation as an Alias is important. 
It can allow the combination of two or more fields. 
It is also useful when the column names are big or not readable, which may or may not affect JSON size and speed depending on network and memory.

**Reply by rijkvanzanten** [reply]
Sure! That's a great use for _aliases_ in general (especially combined with "computational fields" (eg sum/avg etc)), but doesn't really explain why _nested_ data should be under a different key from the source key 🙂

**Reply by philleepflorence** [reply]
I agree...trying to simplify or qualify the request may help.

The concept of Alias or Functions should provide a solution, if I understand the request correctly. 
So while the type of field might be Object and dynamic in the JSON, you can request the field ID as an INT using Alias in the query.

I learned the hard way to try to get simple requests from clients that would give technical requests that may confuse me or my colleagues and not provide the solution they think it might.

**Reply by apstanisic** [reply]
It has nothing to do with `_id`, I just used that to try to example better. Keys could easily be `user` to `user_data`, for example.

In Dart, there is no union type so field can either be `String user` or `DirectusUser user`, there is no union type. So I can't write `DirectusUser | String user`. That forces me to use `Object`, and that removes type safety. 
For example folder's parent is type of `Object`, because I can't write `String | DirectusFolder parent`.

I've thought about it more, and I think that it's easier to set type to `Object`, and force user to use runtime checking that to maintain additional parsing and converting for both API and SDK.

**Reply by rijkvanzanten** [reply]
@apstanisic But at the time you make the request, you already know the data structure you expect right? Wouldn't you be able to type it in runtime at that point?

Edit: 

> and force user to use runtime checking that to maintain additional parsing and converting for both API and SDK.

Ah I see, this is re: the SDK 🙂 Yeah, in that case I'd force the user to type it, as there's no way of knowing what the exact structure is in build-time (as permissions/custom fields etc alter the returned structure)

**Reply by apstanisic** [reply]
SDK returns `Map` by default for items, but has typed classes for system collections.
There would need to be a lot of checking before converting `Map` to system class, so it's probably best to set type as `Object`.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
